### ISSUE TITLE
Add cache control configuration option

### ISSUE THREAD

**Issue by mellis481** [single-spa/import-map-deployer#87]
This issue is a continuation of the discussion from [this](https://github.com/single-spa/import-map-deployer/issues/50#issuecomment-726263780) item regarding adding a new configuration option to be able to set the cache-control header for the importmap.json file.  The default is `public, must-revalidate, max-age=0` which results in the file being cached at cloud edge locations for an indeterminate amount of time.

In the linked issue, there was talk of making it a new configuration item.  After reviewing the documentation again, however, I'm wondering if it's currently possible using the `s3.putObject` property.  So... including something like this in the import-map-deployer's `config.json`:

```
{
  "s3": {
    "putObject": {
      "CacheControl": "no-store"
    }
  }
}
```
@joeldenning Would this work or is there a need to add something new?  Based on [how](https://github.com/single-spa/import-map-deployer/blob/master/src/io-methods/s3.js#L63) it was coded, it looks like it would work.

**Comment by mellis481** [comment]
It appears that this setting does NOT result in the cache-control header of the importmap.json file be set with an overridden value.

In the thread I linked above, I believe an assertion was made that the import-map-deployer will only be able to set/update the cache-control header value of the importmap.json file if the importmap.json file is living with the import-map-deployer service.  That is not how my system is setup:  the import-map-deployer service is running in an EC2 instance and the importmap.json file is hosted in an S3 bucket.  And yet every time the import-map-deployer is called - even with the `s3.putObject` configuration described above - the cache-control value for the importmap.json file keeps being reset to `public, must-revalidate, max-age=0`.

**Comment by joeldenning** [comment]
> I'm wondering if it's currently possible using the s3.putObject property.

No this doesn't work - not sure where you got the idea to try it. We don't pass the import-map-deployer config straight through the the AWS SDK.

I think the best approach would be to add a new configuration item called `cacheControl`. That way we could respect the configuration for other backends, too, not just for S3.

```json
{
  "urlSafeList": ["https://unpkg.com/", "https://my-organization-cdn.com/"],
  "username": "admin",
  "password": "1234",
  "manifestFormat": "importmap|sofe",
  "cacheControl": "public, must-revalidate, max-age=0",
  "locations": {
    "default": "import-map.json",
    "prod": "s3://cdn.canopytax.com/import-map.json",
    "test": "import-map-test.json"
  }
}
```

**Comment by mellis481** [comment]
@joeldenning Glancing at the s3.js file, it looked like the following code would use the `s3.putObject` config to override the default values when writing the importmap.json file:

```
exports.writeManifest = function (filePath, data) {
  const jsonPromise = new Promise(function (resolve, reject) {
    const file = parseFilePath(filePath);
    s3.putObject(
      {
        Bucket: file.bucket,
        Key: file.key,
        Body: data,
        ContentType: "application/importmap+json",
        CacheControl: "public, must-revalidate, max-age=0",
        ACL: "public-read",
        ...s3PutObjectConfig,
      },
      function (err) {
        if (err) reject(err);
        else resolve();
      }
    );
  });
```

I guess I'm not reading it correctly?

**Comment by dgreene1** [comment]
@joeldenning would you be comfortable with us writing validation and unit test testing that ensures the following?

- a user can only set `CacheControl` to either
  - `"no-store"`
  - a string ending in `"must-revalidate, max-age=0"`
  
I feel that it's important to do this since (from my perspective), the whole point of the import map is that it's fetched fresh anytime an update to it is made.

**Comment by mellis481** [comment]
@dgreene1 There are enough options with the [cache-control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) header that I'm not sure the second validation item (a string ending in `"must-revalidate, max-age=0"`) is reasonable unless we would want to be very stringent.  Order doesn't matter for the comma-delimited cache-control directives FWIW.  Also, eg. someone could want to use `s-maxage` instead of `max-age`.

If we want to be stringent, we could say the only allowed values are `no-store`, `public, must-revalidate, max-age=0`, and maybe `no-cache`?  Going the other way (which is what my PR currently has), we could not validate at all and put the onus on the consumer to ensure the cache-control header is correct.

**Comment by joeldenning** [comment]
> I guess I'm not reading it correctly? Is this not where the import map is written to S3? I do understand that this would only work for AWS.

@mellis481  it looks like i misremembered the code we had in s3 - yes it looks like we do pass through the s3 options. Not sure why that didn't work for you. I still think the `cacheControl` option is probably best as a top level config, though, so that it works with all io-methods. I've merged that PR and will release a new version shortly.

> would you be comfortable with us writing validation and unit test testing that ensures the following?
> a user can only set CacheControl to either
> "no-store"
> a string ending in "must-revalidate, max-age=0"

@dgreene1 no, I don't think we can safelist all valid cache control headers, as there are dozens (hundreds?) of organizations using import-map-deployer with myriad use cases. Cache busting is notoriously tricky, and I don't think it makes sense for this project to assert that it knows better than its users about how to do it. Providing a good default is important, which I think we are doing. I agree with what @mellis481 said here:

> There are enough options with the cache-control header that I'm not sure the second validation item (a string ending in "must-revalidate, max-age=0") is reasonable unless we would want to be very stringent

**Comment by joeldenning** [comment]
The `cacheControl` option was implemented in https://github.com/single-spa/import-map-deployer/pull/89, and fixed in #90. I have released it in https://github.com/single-spa/import-map-deployer/releases/tag/v3.2.0

**Comment by mellis481** [comment]
> @mellis481 it looks like i misremembered the code we had in s3 - yes it looks like we do pass through the s3 options. Not sure why that didn't work for you.

I actually think I had a syntax issue. It should work and FWIW it would actually override the `cacheControl` config value if you also supplied `s3.putObject.CacheControl` (for AWS only of course). Not sure if this is a concern. If so, we'd just want to move the CacheControl set line in s3.js below the s3Object spreading.
